server/src/main/java/org/opensearch/index/engine/NRTReplicationEngine.java

```diff
@@ -0,0 +1,330 @@
+        addDeleteOperationToTranslog(delete, deleteResult);
+        deleteResult.setTook(System.nanoTime() - delete.startTime());
+        deleteResult.freeze();
+        localCheckpointTracker.markSeqNoAsProcessed(deleteResult.getSeqNo());
+        return deleteResult;
+    }
+
+    @Override
+    public NoOpResult noOp(NoOp noOp) throws IOException {
+        NoOpResult noOpResult = new NoOpResult(noOp.primaryTerm(), noOp.seqNo());
+        final Translog.Location location = translog.add(new Translog.NoOp(noOp.seqNo(), noOp.primaryTerm(), noOp.reason()));
+        noOpResult.setTranslogLocation(location);
+        noOpResult.setTook(System.nanoTime() - noOp.startTime());
+        noOpResult.freeze();
+        localCheckpointTracker.markSeqNoAsProcessed(noOpResult.getSeqNo());
+        return noOpResult;
+    }
+
+    @Override
+    public GetResult get(Get get, BiFunction<String, SearcherScope, Searcher> searcherFactory) throws EngineException {
+        return getFromSearcher(get, searcherFactory, SearcherScope.EXTERNAL);
+    }
+
+    @Override
+    protected ReferenceManager<OpenSearchDirectoryReader> getReferenceManager(SearcherScope scope) {
+        return readerManager;
+    }
+
+    @Override
+    public Closeable acquireHistoryRetentionLock() {
+        throw new UnsupportedOperationException("Not implemented");
+    }
+
+    @Override
+    public Translog.Snapshot newChangesSnapshot(
+        String source,
+        long fromSeqNo,
+        long toSeqNo,
+        boolean requiredFullRange,
+        boolean accurateCount
+    ) throws IOException {
+        throw new UnsupportedOperationException("Not implemented");
+    }
+
+    @Override
+    public int countNumberOfHistoryOperations(String source, long fromSeqNo, long toSeqNumber) throws IOException {
+        return 0;
+    }
+
+    @Override
+    public boolean hasCompleteOperationHistory(String reason, long startingSeqNo) {
+        return false;
+    }
+
+    @Override
+    public long getMinRetainedSeqNo() {
+        return localCheckpointTracker.getProcessedCheckpoint();
+    }
+
+    @Override
+    public long getPersistedLocalCheckpoint() {
+        return localCheckpointTracker.getPersistedCheckpoint();
+    }
+
+    public long getProcessedLocalCheckpoint() {
+        return localCheckpointTracker.getProcessedCheckpoint();
+    }
+
+    @Override
+    public SeqNoStats getSeqNoStats(long globalCheckpoint) {
+        return localCheckpointTracker.getStats(globalCheckpoint);
+    }
+
+    @Override
+    protected LocalCheckpointTracker getLocalCheckpointTracker() {
+        return localCheckpointTracker;
+    }
+
+    @Override
+    public long getIndexBufferRAMBytesUsed() {
+        return 0;
+    }
+
+    @Override
+    public List<Segment> segments(boolean verbose) {
+        return Arrays.asList(getSegmentInfo(lastCommittedSegmentInfos, verbose));
+    }
+
+    @Override
+    public void refresh(String source) throws EngineException {}
+
```

Comment:
Is this expected to be implemented in future?

Reply:
No it wouldn't.  Refreshes of this engine will only happen after it has received new segments from an external source.

I decided to noOp this for now along with flush() over throwing an Exception to not have extra conditional checks inside IndexShard.

Reply:
writeIndexingBuffer and skipTranslogRecovery both fall under the same case.  I decided to throw the UnsupportedOperationException for methods with a return type, and noOp the others.
I could add them everywhere, but this would mean adding more checks from IndexShard.  I'd prefer to defer that for now and take a closer look at potentially subclassing IndexShard when those conditions are added.